src/main/java/randoop/output/FailingTestFilter.java

```diff
@@ -0,0 +1,277 @@
+      } catch (ProcessException e) {
+        throw new BugInRandoopException("Error filtering regression tests", e);
+      }
+
+      if (status.exitStatus == 0) {
+        passing = true;
+      } else {
+        classSource = commentFailingAssertions(classname, classSource, status);
+      }
+      pass++;
+    }
+    return javaFileWriter.writeClassCode(packageName, classname, classSource);
+  }
+
+  @Override
+  public File writeUnmodifiedClassLines(
+      String packageName, String classname, String[] sourceLines) {
+    return javaFileWriter.writeUnmodifiedClassLines(packageName, classname, sourceLines);
+  }
+
+  @Override
+  public File writeUnmodifiedClassCode(String packageName, String classname, String javaCode) {
+    return javaFileWriter.writeClassCode(packageName, classname, javaCode);
+  }
+
+  /**
+   * Uses the failures in the {@code status} from running JUnit with {@code javaCode} to identify
+   * lines with failing assertions and replaces them with a line comment with the assertion text
+   *
+   * @param classname the name of the test class
+   * @param javaCode the source code for the test class, each assertion must be on its own line
+   * @param status the {@link RunCommand.Status} for running the test with JUnit
+   * @return the class source edited so that failing assertions are replaced by line comments
+   * @throws BugInRandoopException if {@code status} contains output for a failure not involving a
+   *     Randoop generated test method
+   */
+  private String commentFailingAssertions(
+      String classname, String javaCode, RunCommand.Status status) {
+    // JUnit4 writes to standard out, check this doesn't change.
+    assert status.errorOutputLines.isEmpty()
+        : "Expecting JUnit to write to standard out, but found output on standard error";
+
+    /* Iterator to move through JUnit output. */
+    Iterator<String> lineIterator = status.standardOutputLines.iterator();
+
+    /*
+     * First, find the message that indicates the number of failures in the run.
+     */
+
+    Pair<String, String> failureCountMatch = readUntilMatch(lineIterator, FAILURE_MESSAGE_PATTERN);
+    int totalFailures = Integer.parseInt(failureCountMatch.b);
+    assert totalFailures > 0 : "JUnit has non-zero exit status, but no failure found";
+
+    /*
+     * Then read the rest of the file to find each failure.
+     * The standard runner gives a numbered list with each entry matching the following pattern:
+     */
+    Pattern failureHeaderPattern =
+        Pattern.compile("\\d+\\)\\s+(" + ID_STRING + ")\\(" + ID_STRING + "\\)");
+
+    /*
+     * Split Java code text so that we can match the line number for the assertion with the code.
+     * Use same line break as used to write test class file.
+     */
+    String[] javaCodeLines = javaCode.split(Globals.lineSep);
+
+    int failureCount = 0;
+    while (failureCount < totalFailures) {
```

Comment:
Why isn't this a `for` loop?  That brings all the control structure into 1 statement; currently it's on 2 statements here and on 1 more statement that is 43 lines away from these.